framework/client.py

```diff
@@ -79,30 +82,45 @@ def handleSimulationStop(self, packet):
       await self.handler()
 
   async def handler(self):
-      listener_task = asyncio.ensure_future(self.ws.recv())
-      producer_task = asyncio.ensure_future(self.fetchMessage())
-      done, pending = await asyncio.wait([listener_task, producer_task],
-                                         return_when=asyncio.FIRST_COMPLETED)
-
-      if producer_task in done:
-        packet = producer_task.result()
-        await self.send_packet(packet)
-      else:
-        producer_task.cancel()
-
-      #if the connection is dead we will reach this point
-      if not self.alive:
-        self.active = False
-        for fut in pending:
-          fut.cancel()
-        return
-
-      if listener_task in done:
-        message = listener_task.result()
-        packet = json.loads(message)
-        loop.run_in_executor(None, self.onMessage, packet)
-      else:
-        listener_task.cancel()
+    listener_task = asyncio.ensure_future(self.ws.recv())
+    producer_task = asyncio.ensure_future(self.fetchMessage())
+    done, pending = await asyncio.wait([listener_task, producer_task],
+                                       return_when=asyncio.FIRST_COMPLETED)
+
+    if producer_task in done:
+      packet = producer_task.result()
+      await self.send_packet(packet)
+    else:
+      producer_task.cancel()
+
+    #if the connection is dead we will reach this point
+    if not self.alive:
+      self.active = False
+      for fut in pending:
+        fut.cancel()
+      return
+
+    if listener_task in done:
+      message = listener_task.result()
+      packet = json.loads(message)
+      loop.run_in_executor(None, self.onMessage, packet)
+    else:
+      listener_task.cancel()
+
+  def failsafe(self, error):
```

Comment:
The `failsafe` function sets `alive` to false, which will shut down the framework. This will cause the socket connection it has to close, which will then cause the server to respond appropriately.